app/node/distort/cornerpin/cornerpindistortnode.cpp

```diff
@@ -0,0 +1,237 @@
+  switch (value) {
+    case 0: // Top left
+      return QPointF(row[kTopLeftInput].data().value<QVector2D>().x(),
+                     row[kTopLeftInput].data().value<QVector2D>().y());
+      break;
+    case 1: // Top right
+      return QPointF(resolution.x() + row[kTopRightInput].data().value<QVector2D>().x(),
+                     row[kTopRightInput].data().value<QVector2D>().y());
+      break;
+    case 2: // Bottom right
+      return QPointF(resolution.x() + row[kBottomRightInput].data().value<QVector2D>().x(),
+                     resolution.y() + row[kBottomRightInput].data().value<QVector2D>().y());
+      break;
+    case 3: //Bottom left
+      return QPointF(row[kBottomLeftInput].data().value<QVector2D>().x(),
+                     row[kBottomLeftInput].data().value<QVector2D>().y() + resolution.y());
+      break;
+    default: // We should never get here
+      return QPointF();
+  }
+}
+
+void CornerPinDistortNode::DrawGizmos(const NodeValueRow &row, const NodeGlobals &globals, QPainter *p)
+{
+  const QVector2D &resolution = globals.resolution();
+
+  const double handle_radius = GetGizmoHandleRadius(p->transform());
+
+  p->setPen(QPen(Qt::white, 0));
+
+  QPointF top_left = ValueToPixel(0, row, resolution);
+  QPointF top_right = ValueToPixel(1, row, resolution);
+  QPointF bottom_right = ValueToPixel(2, row, resolution);
+  QPointF bottom_left = ValueToPixel(3, row, resolution);
+
+  // Add the correct offset to each slider
+  SetInputProperty(kTopLeftInput, QStringLiteral("offset"), QVector2D(0.0, 0.0));
+  SetInputProperty(kTopRightInput, QStringLiteral("offset"), QVector2D(resolution.x() , 0.0));
+  SetInputProperty(kBottomRightInput, QStringLiteral("offset"), resolution);
+  SetInputProperty(kBottomLeftInput, QStringLiteral("offset"), QVector2D(0.0, resolution.y()));
+
+  // Draw bounding box
+  p->drawLine(QLineF(top_left, top_right));
+  p->drawLine(QLineF(top_right, bottom_right));
+  p->drawLine(QLineF(bottom_right, bottom_left));
+  p->drawLine(QLineF(bottom_left, top_left));
+
+  // Create handles
+  gizmo_resize_handle_[0] = CreateGizmoHandleRect(top_left, handle_radius);
+  gizmo_resize_handle_[1] = CreateGizmoHandleRect(top_right, handle_radius);
+  gizmo_resize_handle_[2] = CreateGizmoHandleRect(bottom_right, handle_radius);
+  gizmo_resize_handle_[3] = CreateGizmoHandleRect(bottom_left, handle_radius);
+
+  // Draw handles
+  DrawAndExpandGizmoHandles(p, handle_radius, gizmo_resize_handle_, kGizmoCornerCount);
+}
+
+bool CornerPinDistortNode::GizmoPress(const NodeValueRow &row, const NodeGlobals &globals, const QPointF &p)
+{
+  bool gizmo_active[kGizmoCornerCount] = {false};
+
+  for (int i = 0; i < kGizmoCornerCount; i++) {
+    gizmo_active[i] = gizmo_resize_handle_[i].contains(p);
+
+    if (gizmo_active[i]) {
+      gizmo_drag_start_ = p;
+      gizmo_res_ = globals.resolution();
+      gizmo_drag_ = i;
+
+      switch (i) {
+        case 0:
+          gizmo_start_ = row[kTopLeftInput].data();
```

Comment:
NodeInputDragger now has a `GetStartValue` function which can be used rather than having to cache the start value in `gizmo_start_`. Note that GetStartValue will be a double rather than a QVector2D since it's only one track's value.